### ISSUE TITLE
🚀 Feature: Allow notification feed filtering by custom payload data

### ISSUE THREAD

**Issue by chauandrew** [novuhq/novu#3466]
### 🔖 Feature description

At the moment, we can query notification meessages/feeds by `page`, `feedId`, `seen`, `read`, `subscriberId`, `channels`, etc. It would be helpful to query by custom data in `messages.payload` 

### 🎤 Why is this feature needed ?

I work on a web app where we are trying to migrate our existing notification systems to Novu. One feature of our existing notification system is that we are able to filter notifications by their 'component type', or a subsection of the app. As we transition to Novu, we can store this custom data within a notification payload, but we run into compatibility issues since we are unable to filter by these features.

### ✌️ How do you aim to achieve this?

I would like to expand the `GET /v1/notifications` api to take custom data as input.

Specifically, I think it will involve the following:

- Add a new parameter to`@novu/dal/src/repositories/messages/message.repository.ts > MessageRepository > getFilterQueryForMessage()` to take custom payload data as input
- Update `@novu/api/src/app/widgets/usecases/get-notifications-feed` and the corresponding `StoreQuery` object to take custom payload data 

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by chauandrew** [comment]
Good catch I completely missed [that](https://github.com/novuhq/novu/issues/3230). Seems like this ticket makes sense as an API change, and the other front end dependency once this is finished. But I'm new to the codebase, so open to feedback / other suggestions!

**Comment by chauandrew** [comment]
Can I work on this?

**Comment by jainpawan21** [comment]
cc:- @scopsy

**Comment by scopsy** [comment]
Sure thing @chauandrew!

**Comment by chauandrew** [comment]
Hi @scopsy - sorry about this, recently learned that I'm not supposed to be contributing to any open source projects. Don't think I can take this on.  But if I'm able to get permission in the future I'll try to contribute! 

Thanks!

**Comment by gitstart** [comment]
@scopsy I would like to pick this up

**Comment by thenbe** [comment]
I also have similar use case:

Say you produce an invoice at the beginning of the month.

On day 1, you create a new invoice and send a reminder to the user.
On day 2, the user has not paid, so you send another reminder.
On day 3, the user has not paid, so you send another reminder.
...and so on.

Then, you want to view all the reminders you have sent for this particular invoice. Is there a way to view/paginate those reminders? How can this be done?

### Workaround:

Use the `invoice_id` as the `subscriber_id`. Then, use the existing search functionality to search by `subscriber_id`.

```bash
curl https://api.novu.co/v1/notifications?search=${INVOICE_ID}
```

### Proposed solution:

1. Define a custom_field of `"invoice_id": 123` when sending a notification.

```typescript
await this.novu.trigger('TAG', {
	metadata: {
		invoice_id: 123,
	},
	// ...
});
```

2. Then search for notifications where `"invoice_id" === "123"`. This fetches the desired result and allows for pagination.

```bash
curl https://api.novu.co/v1/notifications?metadata_invoice_id=123
```

### Prior art:

This is the pattern used by the Postmark API:

1. Define a metadata `custom_field = "foo"` when [sending](https://postmarkapp.com/developer/api/templates-api#email-with-template) a message.
1. Afterwards, [search](https://postmarkapp.com/developer/api/messages-api#outbound-message-search) by for messages that have the metadata `custom_field = "foo"`.

**Comment by LetItRock** [comment]
hey @chauandrew 👋 the functionality will be implemented in the PR #3939 ;)
